document/JS.bs

```diff
@@ -0,0 +1,935 @@
+
+typedef record&lt;DOMString, InstanceExportsMap&gt; InstanceImportsMap;
+
+dictionary WebAssemblyInstantiatedSource {
+   required Module module;
+   required Instance instance;
+};
+
+[Exposed=(Window,Worker)]
+namespace WebAssembly {
+  boolean validate(BufferSource bytes);
+  Promise&lt;Module> compile(BufferSource bytes);
+
+  Promise&lt;WebAssemblyInstantiatedSource> instantiate(
+      BufferSource bytes, optional InstanceImportsMap importObject);
+
+  Promise&lt;Instance> instantiate(
+      Module moduleObject, optional InstanceImportsMap importObject);
+};
+</pre>
+
+<!--
+Should we include notes describing what the functions do, as the HTML spec does? It could look like this:
+
+Note:
+  WebAssembly.validate(|bytes|) synchronously validates bytes of WebAssembly, returning true if the validation was successful.
+  WebAssembly.compile(|bytes|) asynchronously validates and complies bytes of WebAssembly into a Module.
+  WebAssembly.instantiate(|bytes|, |importObject|) asynchronously compiles and instantiates a WebAssembly module from bytes of source.
+  The WebAssembly.instantiate(|moduleObject|, |importObject|) asynchronously instantiates a compiled module.
+-->
+
+<div algorithm>
+  A WebAssembly module AST |module| is said to be <dfn>valid</dfn> if:
+    1. [=validate_module=](_module_) is not [=error=], and
+    1. The UTF-8 encoding of all <code>name</code> productions in the binary format, including in a custom <code>name</code> section if present, is valid.
+</div>
+
+<div algorithm>
+  The <dfn method for="WebAssembly">validate(bytes)</dfn> method, when invoked, performs the following steps:
+    1. Let |module| be [=decode_module=](_bytes_). If |module| is [=error=], return false.
+    1. If |module| is not [=valid=], return false.
+    1. Return true.
+</div>
+
+<div algorithm>
+  To <dfn>compile a WebAssembly module</dfn> from a {{BufferSource}} |bytes|, perform the following steps:
+    1. Let |module| be [=decode_module=](_bytes_). If |module| is [=error=], return the string "CompileError".
+    1. If |module| is not [=valid=], return the string "CompileError"
+    1. Return |module|
+</div>
+
+<div algorithm>
+  To <dfn>construct a WebAssembly module object</dfn> from a module |module|, perform the following steps:
+
+    1. Let |moduleObject| be a new {{Module}} object.
+    1. Set |moduleObject|.\[[Module]] to |module|.
+    1. Set |moduleObject|.\[[Bytes]] to |bytes|.
+    1. Return |moduleObject|.
+</div>
+
+<div algorithm>
+  To <dfn>asynchronously compile a WebAssembly module</dfn> from a {{BufferSource}} |bytes|, with the promise |promise|, using optional [=task source=] |taskSource|, perform the following steps:
+
+    1. In parallel, [=compile a WebAssembly module|compile the WebAssembly module=] |bytes| and store the result as |module|.
+    1. When the above operation completes, queue a task to perform the following steps. If |taskSource| was provided, queue the task on that task source.
```

Comment:
"queue a task" should be linked. Specifying the task source is not optional.

Reply:
I'm not sure which task source to specify. HTML seems to queue a lot of tasks without specifying a task source. Web.bs specifies the networking task source, but I'm not sure that makes sense here. I'd like to have wording here which would, in theory, let Node.js use its task queues for a compliant implementation; this is hoped to not depend too tightly on HTML.

Reply:
When in doubt, either make up your own task source, or default to the DOM manipulation task source. Or go with the networking task source for consistency, I suppose.

Neither will pose an issue for Node.js. There's nothing special about the names of the task sources, only their contents; we could have called them task sources 1...N.